### PR TITLE
add exec to ui5 controls to boost performance for data retrieval from many ui5 child controls

### PR DESCRIPTION
**Pull request by philippthiele**
Note: documentation is still missing, please let me know what you think of the PR and whether something can be improved, before I start documentation efforts. Thanks!

image

### CODE REVIEW

**Comment by vobu** [comment]
🚀 you're clearly on a 👨‍💻 spree here @philippthiele 🔋 ¡excellente!
I really dig the idea of iterating over a massive amount of `aggregation` children of a control w/o any burden of browser-/Node.js-scope change overhead! 
Let me think some on the way the `evalOnControl()` API is surfaced though - will need to form an opinion and would also like to get the take on this from @Siolto, @monavari-lebrecht, @dominikfeininger and @ArnaudBuchholz. Colleagues: discuss, please! 😄

**Comment by dominikfeininger** [comment]
Hi, nice feature. Good Idea !

Maybe a name like `executeOnControl` of even `execute`/ `exec` ala. Node could be a possibility.

I added some performance testing to your test file. Resulting in roughly bisection of execution time. 👍 for that.

**Comment by philippthiele** [comment]
@dominikfeininger let me know which name fits best to your conventions, I will adapt or you can also do yourself, as far as I know I have enabled edits for maintainers 😊

**Comment by philippthiele** [comment]
Hi @vobu @dominikfeininger ,
do you feel except from aligning on the naming there is something else missing I could contribute still?
Thanks!

**Comment by JanWerder** [comment]
This would help a lot, would be great if this could be merged. 👍

File: examples/ui5-js-app/package.json
Comment (Siolto) [inline examples/ui5-js-app/package.json:23]: ui5 tooling v3 requires at least node version 16. Currently we still need to support node 14 that's why our tests are failing

File: examples/ui5-js-app/package.json
Comment (Siolto) [inline examples/ui5-js-app/package.json:38]: why do wee need these dependencies?

File: package.json
Comment (Siolto) [inline package.json:79]: same here we still test wdi5 with node 14 as well

**Review by Siolto — CHANGES_REQUESTED** [review]
some node version problems rest looks good. @vobu @dominikfeininger we still want to support node 14, right?

**Comment by Siolto** [comment]
Hi @philippthiele,

great work and nice idea! Looks like this could boost up test execution quite a lot. From my point of view technically this looks fine. Just some minor remarks from my side. Documentation, fixing the tests and naming alignment is open I guess.

File: examples/ui5-js-app/package.json
Comment (vobu) [reply examples/ui5-js-app/package.json:23]: agreed, let's put this on a separate PR to both drop node14 and upgrade to ui5-tooling 3

File: examples/ui5-js-app/webapp/test/e2e/evalOnControl.test.js
Comment (vobu) [inline examples/ui5-js-app/webapp/test/e2e/evalOnControl.test.js:60]: similar to my suggestion above:
```js
await list.evalOnControl( getItems().map( item => item.getTitle() )
```

Comment (vobu) [reply examples/ui5-js-app/webapp/test/e2e/evalOnControl.test.js:60]: this would even allow to dynamically change the return type, à la
```js
// have an object returned with an embedded array
await list.evalOnControl(  { getItems().map( item => item.getTitle() } )
```

File: examples/ui5-js-app/webapp/test/e2e/evalOnControl.test.js
Comment (vobu) [inline examples/ui5-js-app/webapp/test/e2e/evalOnControl.test.js:101]: just for completenes' sake:
```js
browser.asControl(listSelector).evalOnControl(
  {
         tableTitle: getHeaderText(),
         peopleListNames: getItems().map((item) => item.getTitle())
  }
)
```

**Comment by philippthiele** [comment]
@vobu 
Regarding dropping the function:
- I would not drop it completely, but make both options available. In a function I might do much more than just access the functions of the control, options are limitless.
- I technically do not see how to implement the suggestion though, if I pass it like suggested it will just be executed directly and not on browser side. Can you elaborate a bit? Then I will be happy to implement it.

Rename method to exec: done

Documentation: Will do when everything is clarified and implemented 😊

**Comment by vobu** [comment]
> * I would not drop it completely, but make both options available. In a function I might do much more than just access the functions of the control, options are limitless.

ah, of course, didn't have this on the radar! 
`exec( function () { /** stuff */ })` allows to run `stuff` not limited to the retrieved Control → leave it as is then of course!
btw: could you please add a test to make sure that a lambda/arrow fn à la `exec( () => { /** stuff */ })` would also work?

> Rename method to exec: done

👍 

> Documentation: Will do when everything is clarified and implemented 😊

👍 

please also don't forget to downgrade `ui5-tooling` to v2 to re-add Node 14 support

**Comment by philippthiele** [comment]
@vobu I have update the chromedriver in the package lock which fixed a few tests, rest does not seem to be related to a package or the changes in this PR though. Can you help check what is wrong?

**Comment by vobu** [comment]
> @vobu I have update the chromedriver in the package lock which fixed a few tests, rest does not seem to be related to a package or the changes in this PR though. Can you help check what is wrong?

sure, I'll peak and see what I can fix. Let me get back to this as soon as I can.

> Also: I have extended the functionality to accepts arguments as well, as we were talking about anything being possible with a function, passing additional arguments makes sense I would think 😊

¡excellente!

> Reminder for me: I should check whether propagating errors is possible, to make debugging easier when a browser side error occurs, like calling a non-existent function like "button.getTex()"

absolutely, nothing beats DevX improvements.

**Comment by philippthiele** [comment]
@vobu 
Errors are now logged to console, couldn't really verify the log in the test, only manually visually. Verified that the return value is null in this case:
https://github.com/ui5-community/wdi5/pull/456/commits/dddb3b21b0ea1a7ff80ecb2afa56450d55284d56
image

Next step will be documentation.

**Comment by philippthiele** [comment]
@vobu documentation is added 😊 Finishing line is in sight.
Only auth test failing, I can't see any relation to the PR.

Please let me know whether something is missing/can be improved!

File: examples/ui5-js-app/webapp/test/e2e/exec.test.js
Comment (vobu) [inline examples/ui5-js-app/webapp/test/e2e/exec.test.js:18]: optional enhancement: you can validate error messages on the console in the Node.js scope with the help of `sinon`.  
See https://github.com/ui5-community/wdi5/blob/16d69841b7c5d597c6d9712b3c64207cca750ebc/examples/ui5-js-app/webapp/test/e2e/error-logging.test.js#L26

File: examples/ui5-js-app/webapp/test/e2e/exec.test.js
Comment (vobu) [inline examples/ui5-js-app/webapp/test/e2e/exec.test.js:65]: small typo

```suggestion
    it("execute function browserside on button to get its text with fluent async api, basic return type", async () => {
```

**Review by vobu — APPROVED** [review]
this. is. so. cool.
there's a small typo (`suggest`ed a change). 
and i left a hint how to optional test for error messages relayed from browser- to Node-scope, as we have this in a test already.
I'm stoked to see this go in once it has the +1 from Dominik and Simon

**Review by dominikfeininger — APPROVED** [review]

**Review by Siolto — APPROVED** [review]

**Comment by vobu** [comment]
@philippthiele we'll of course announce the new version with this excellent feature, but would you also care to blog about it on blogs.sap.com? Pretty please? 😸

**Comment by philippthiele** [comment]
Thanks to you all for reviewing and merging 😊

@vobu Here are the suggested changes for the test, thanks for that, learned something new: https://github.com/ui5-community/wdi5/pull/466
I will try to write a blog post this week, will be on vacation the next, so it could be time-wise hard to manage. Will let you know by the end of the week.

**Comment by philippthiele** [comment]
@vobu I will not be able to produce an adequate blog post before my vacation, I will be happy to pick up the topci afterwards 😊

**Comment by vobu** [comment]
all good! enjoy your time off and recharge properly! looking forward to reading your blog post after your vacation 😄

**Pull request merged**
